src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -0,0 +1,551 @@
+    filter: u64,
+    /// Mask value to be applied.
+    #[serde(deserialize_with = "deserialize_u64_from_str")]
+    value: u64,
+}
+
+/// Generic struct for addressing registers
+/// that are non-CPUID based.
+#[derive(Debug, Deserialize, Eq, PartialEq)]
+pub struct RegisterId {
+    #[serde(deserialize_with = "deserialize_u64_from_str")]
+    value: u64,
+}
+
+/// Errors thrown while configuring templates.
+#[derive(Debug, thiserror::Error, PartialEq, Eq)]
+pub enum Error {
+    /// Failure in processing the CPUID configuration.
+    #[error("Error deserializing CPUID")]
+    Cpuid,
+    /// Unknown failure in processing the CPU template.
+    #[error("Internal error processing CPU template")]
+    Internal,
+}
+
+/// Guest config sub-module specifically useful for
+/// config templates.
+#[cfg(target_arch = "x86_64")]
+pub mod x86_64 {
+    use std::collections::{BTreeMap, HashMap};
+    use std::str::FromStr;
+
+    use serde::de::Error as SerdeError;
+    use serde::{Deserialize, Deserializer};
+
+    use crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags;
+    use crate::guest_config::cpuid::{
+        AmdCpuid, Cpuid, CpuidEntry, CpuidKey, CpuidRegisters, IntelCpuid,
+    };
+    use crate::guest_config::templates::{BitMask, Error, RegisterModifier};
+    use crate::guest_config::x86_64::X86_64CpuConfiguration;
+
+    /// CPUID register enumeration
+    #[allow(unused)]
+    #[allow(missing_docs)]
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub enum CpuidRegister {
+        Eax,
+        Ebx,
+        Ecx,
+        Edx,
+    }
+
+    /// Composite type that holistically provides
+    /// the location of a specific register being used
+    /// in the context of a CPUID tree.
+    #[allow(unused)]
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct CpuidRegisterModifier {
+        /// Leaf value.
+        #[serde(deserialize_with = "deserialize_u32_from_str")]
+        pub leaf: u32,
+        /// Sub-Leaf value.
+        #[serde(deserialize_with = "deserialize_u32_from_str")]
+        pub subleaf: u32,
+        /// KVM feature flags for this leaf-subleaf.
+        #[serde(deserialize_with = "deserialize_kvm_cpuid_flags")]
+        pub flags: crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags,
+        /// CPUID register to be modified by the bitmask.
+        #[serde(deserialize_with = "deserialize_cpuid_register")]
+        pub register: CpuidRegister,
+        /// Mask to be applied to register's value at the address
+        /// provided.
+        pub mask: BitMask,
+    }
+
+    /// Wrapper type to containing x86_64 CPU config modifiers.
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct X86_64CpuTemplate {
+        /// Modifiers for CPUID configuration.
+        #[serde(default)]
+        pub cpuid_modifiers: Vec<CpuidRegisterModifier>,
+        /// Modifiers for model specific registers.
+        #[serde(default)]
+        pub msr_modifiers: Vec<RegisterModifier>,
```

Comment:
AFAIK, as rdmsr takes the ECX register (32-bit), while the return value is EDX:EAX (64-bit), this might not work.
https://www.felixcloutier.com/x86/rdmsr